### PR TITLE
Stop copying LogicalPlan and Exprs in `RewriteDisjunctivePredicate`

### PR DESCRIPTION
**Pull request by rohitrastogi**
## Which issue does this PR close?
Part of https://github.com/apache/datafusion/issues/9637
Closes https://github.com/apache/datafusion/issues/10213

### CODE REVIEW

File: datafusion/optimizer/src/rewrite_disjunctive_predicate.rs [hunk 1]
```diff
@@ -56,29 +58,34 @@ use datafusion_expr::{Expr, LogicalPlan, Operator};
 ///
 /// ```sql
 /// where
-///   p_partkey = l_partkey
```
Comment (rohitrastogi) [inline datafusion/optimizer/src/rewrite_disjunctive_predicate.rs:59]: First SQL block did not show query before rewrite.

File: datafusion/optimizer/src/rewrite_disjunctive_predicate.rs [hunk 2]
```diff
@@ -305,31 +316,30 @@ fn delete_duplicate_predicates(or_predicates: &[Predicate]) -> Predicate {
...
-                let mut new_args = (*args).clone();
-                new_args.retain(|expr| !exist_exprs.contains(expr));
-                if !new_args.is_empty() {
-                    if new_args.len() == 1 {
-                        new_or_predicates.push(new_args[0].clone());
+            Predicate::And { mut args } => {
+                args.retain(|expr| !exist_exprs.contains(expr));
+                if !args.is_empty() {
+                    if args.len() == 1 {
+                        new_or_predicates.push(args.remove(0));
```
Comment (rohitrastogi) [inline datafusion/optimizer/src/rewrite_disjunctive_predicate.rs:332]: I believe remove() should be O(1) here and below as there's just one item in the vector.

**Review by alamb — APPROVED** [review]
This is great @rohitrastogi  -- thank you. A very nice first contribution 🙏

**Review by comphead — APPROVED** [review]
lgtm thanks @rohitrastogi

**Pull request merged**